zyte_common_items/pipelines.py

```diff
@@ -53,3 +55,78 @@ def __init__(self):
 
     def process_item(self, item, spider):
         return ae.downgrade(item)
+
+
+class DropLowProbabilityItemPipeline:
+    """This pipeline drops an item if its probability, defined in the settings,
+    is less than the specified threshold.
+
+    There are two ways that the thresholds can be defined per item type using
+    the ``ITEM_PROBABILITY_THRESHOLDS`` setting:
+
+    With paths to Item classes:
+        ITEM_PROBABILITY_THRESHOLDS = {
+            "zyte_common_items.items.Article": 0.1,
+            "zyte_common_items.items.Product": 0.2
+        }
+
+    With Item classes:
+        from zyte_common_items import Article, Product
+
+        ITEM_PROBABILITY_THRESHOLDS = {
+            Article: 0.1,
+            Product: 0.2
+        }
+    """
+
+    def __init__(self, crawler):
+        self.stats = crawler.stats
+        self.thresholds = {}
+        self.init_thresholds(crawler.spider)
+
+    @classmethod
+    def from_crawler(cls, crawler):
+        return cls(crawler)
+
+    def init_thresholds(self, spider):
+        from scrapy.utils.misc import load_object
+
+        thresholds_settings = spider.settings.get("ITEM_PROBABILITY_THRESHOLDS", {})
+        for item, threshold in thresholds_settings.items():
+            item_type = load_object(item) if isinstance(item, str) else item
+            self.thresholds[item_type] = threshold
+
+    def get_threshold_for_item(self, item, spider):
+        return self.thresholds.get(type(item), DEFAULT_ITEM_PROBABILITY_THRESHOLD)
+
+    def get_item_name(self, item):
+        return item.__class__.__name__
+
+    def process_item(self, item, spider):
+        from scrapy.exceptions import DropItem
+
+        item_name = self.get_item_name(item)
+        item_proba = item.get_probability()
+        threshold = self.get_threshold_for_item(item, spider)
+
+        self.stats.inc_value("drop_low_probability_item/processed", spider=spider)
```

Comment:
I _think_ all these spider=spider can be dropped; I have zero idea why do we have this parameter in StatsCollector :) Probably, it's a leftover from times where Crawler supported multiple spiders.

Reply:
Oh, thanks, TIL!